nextcloud/talk-android
# Fix crash when setting layout for call participants

Fixes #2445, which is a regression introduced in #2390 (specifically, in https://github.com/nextcloud/talk-android/pull/2390/commits/c3f1f6c3a8ff74930b91b5f00c6435186f5ec1e2)

When the layout for a participant is setup and the HPB is not used the avatar is based on the ID of the user, [which is got from the list of known participants in the call](https://github.com/nextcloud/talk-android/blob/f880b4f7787d1349e809d3e0847921aaffe346f8/app/src/main/java/com/nextcloud/talk/activities/CallActivity.java#L2364). However, the list of known participants [is fetched](https://github.com/nextcloud/talk-android/blob/f880b4f7787d1349e809d3e0847921aaffe346f8/app/src/main/java/com/nextcloud/talk/activities/CallActivity.java#L1891) just [before setting up the layout](https://github.com/nextcloud/talk-android/blob/f880b4f7787d1349e809d3e0847921aaffe346f8/app/src/main/java/com/nextcloud/talk/activities/CallActivity.java#L1904). Therefore, when the layout is setup the list of known participants does not contain yet the participant, which ended in a `NullPointerException`.

To prevent that now it is verified if the participant is in the list before getting the user ID, and if not an empty user ID is used. However, that would cause that all users are initially shown as guests. Fortunately the user ID is provided by the signaling message used to create the participant and thus setup the initial layout, so it can be provided to `setupVideoStreamForLayout` to show the proper avatar even if the list of known participants was not fetched yet.

Note that, although highly unlikely, it would be possible that a connection is established with the remote participant before the data for the participants was received. In that case the user ID is provided to `setupVideoStreamForLayout` when the remote participant was initially found, but not when the connection was established, so if the remote participant has her video disabled a guest avatar would be shown due to the missing user ID. This already happened before and it is not introduced (nor fixed) by this pull request.

Although the issues happens when the HPB is not used it would be good to test too with the HPB to ensure that nothing breaks ;-)

## How to test

- Do not setup the HPB
- Ensure that getting the peers will take longer than setting the layout by adding something like `sleep(2);` to the beginning of [`getPeersForCall`](https://github.com/nextcloud/spreed/blob/62f47af2b6143d7c9ff1f24b1ffce2ff6951bc7f/lib/Controller/CallController.php#L69)
- Start a call with the WebUI
- In the Android app, join the call

### Result with this pull request

The app does not crash (and the avatar for the remote participant is shown before establishing a connection)

### Result without this pull request

The app crashes
